src/helics/common/SmallBuffer.cpp

```diff
@@ -0,0 +1,260 @@
+}
+
+SmallBuffer& SmallBuffer::operator=(const SmallBuffer& sb)
+{
+    if (this == &sb) {
+        return *this;
+    }
+    resize(sb.size());
+    std::memcpy(heap, sb.heap, sb.size());
+    return *this;
+}
+
+SmallBuffer& SmallBuffer::operator=(std::string_view val)
+{
+    if (reinterpret_cast<const std::byte*>(val.data()) == heap) {
+        bufferSize = val.size();
+        return *this;
+    }
+    resize(val.size());
+    std::memcpy(heap, val.data(), val.size());
+    return *this;
+}
+
+SmallBuffer& SmallBuffer::operator=(SmallBuffer&& sb) noexcept
+{
+    if (usingAllocatedBuffer) {
+        if (nonOwning) {
+            if (sb.heap == heap) {
+                bufferSize = sb.bufferSize;
+                bufferCapacity = sb.bufferCapacity;
+                return *this;
+            }
+        } else {
+            if (sb.heap == heap) {
+                bufferSize = sb.bufferSize;
+                return *this;
+            }
+            delete[] heap;
+        }
+    }
+    if (sb.usingAllocatedBuffer) {
+        heap = sb.heap;
+        bufferCapacity = sb.bufferCapacity;
+        usingAllocatedBuffer = true;
+        nonOwning = sb.nonOwning;
+    } else {
+        std::memcpy(buffer.data(), sb.heap, sb.bufferSize);
+        usingAllocatedBuffer = false;
+        nonOwning = false;
+        heap = buffer.data();
+        bufferCapacity = 64;
+    }
+    bufferSize = sb.bufferSize;
+    sb.heap = sb.buffer.data();
+    sb.bufferCapacity = 64;
+    sb.bufferSize = 0;
+    sb.usingAllocatedBuffer = false;
+    return *this;
+}
+
+std::byte* SmallBuffer::release()
+{
+    if (!usingAllocatedBuffer) {
+        return nullptr;
+    }
+    auto* released = heap;
+    heap = buffer.data();
+    usingAllocatedBuffer = false;
+    nonOwning = false;
+    bufferCapacity = 64;
+    bufferSize = 0;
+    return released;
+}
+
+void SmallBuffer::moveAssign(void* data, std::size_t size, std::size_t capacity)
+{
+    auto* newHeap = reinterpret_cast<std::byte*>(data);
+    if (usingAllocatedBuffer && !nonOwning) {
+        if (newHeap != heap) {
+            delete[] heap;
+        }
+    }
+    heap = newHeap;
+    bufferCapacity = capacity;
+    bufferSize = size;
+    nonOwning = false;
+    usingAllocatedBuffer = true;
+}
+
+void SmallBuffer::spanAssign(void* data, std::size_t size, std::size_t capacity)
+{
+    auto* newHeap = reinterpret_cast<std::byte*>(data);
+    if (usingAllocatedBuffer && !nonOwning) {
+        if (newHeap == heap) {
+            // if the heaps are the same the only thing to change is the size and capacity
+            // don't change the other characteristics
+            bufferSize = size;
+            bufferCapacity = capacity;
+            return;
+        }
+        delete[] heap;
+    }
+    heap = newHeap;
+    bufferCapacity = capacity;
+    bufferSize = size;
+    nonOwning = true;
+    usingAllocatedBuffer = true;
+}
+
+void SmallBuffer::assign(const void* start, const void* end)
```

Comment:
Does a check to make sure end >= start and throw an exception need to get added? It looks like std::vector's assign method throws a bad_alloc exception in that scenario, but I don't know if that's something in their assign method implementation, or elsewhere in the standard library that results in that behavior.

Reply:
I added the check, end==start isn't a problem but end < start could be.  So I am throwing a std::invalid_argument in that case.

Reply:
So end>=start is okay, with end==start being the "empty" case.

Reply:
In "empty" case described by @nightlark is there a desired behavior if the `operator[]` is used to access a member of the size-zero heap? At present, it is a segfault.

Reply:
I don't think it would be a segfault.  at size 0 it would still be using the internal buffer so would have a minimum capacity of 64, which means anything 0-63 and likely a few more wouldn't be  segfault